Added redis-client (Bash): https://github.com/SomajitDey/redis-client

SomajitDey opened the pull request:
The only existing Bash client featured in https://redis.io/clients#bash, viz. [Redi.sh](https://github.com/crypt1d/redi.sh), is very limited in scope and not very extensible. Although written in Bash, Redi.sh depends unnecessarily on various Unix tools rather than using Bash native functionalities mostly.

The present [client](https://github.com/SomajitDey/redis-client), on the contrary, is written from scratch with a strong focus on using Bash functionalities as much as possible. This adds to its performance and portability. It is also extensible, and currently has full support for any Redis command execution in the request-response mode.  Support for pipelining and push protocol would be added in near future.

The client library also supports creating a Redis session and keeping it alive, instead of creating a new session for every command execution - a feature using which _connection pools_ may be implemented in future. 

Using the client library, a portable Redis command-line-interface (CLI or console) has also been implemented.

oranagra commented:
@itamarhaber @gkorland do you have anything to contribute / argue here?

zuiderkwast commented:
Looks good to me.

@SomajitDey I think you can remove the `"active": true` flag of redi.sh, if you want, since its last commit was 5 years ago.

zuiderkwast commented:
Hello again. I decided to try it out. It's seems to work for simple commands but it's very mature:

```
$ ./redis-cli 
localhost:6379_0$ get foo
bar
localhost:6379_0$ keys *
./redis.bash: eval: line 104: unexpected EOF while looking for matching `)'
localhost:6379_0$ keys (*)
NuLL
localhost:6379_0$ command
replicaof lrem geoadd bitfield_ro geodist move bitop hgetall bitfield swapdb zrevrank lindex flushall sadd hvals hexists xclaim georadiusbymember acl sinter auth xtrim sync linsert discard lolwut zremrangebyscore dbsize hlen rpop brpop exec mget pubsub scard slaveof client unwatch zrangebyscore shutdown zincrby lpop decr zunionstore sscan substr srem zremrangebylex zrange lpushx setnx xreadgroup append getbit bgrewriteaof xrevrange zremrangebyrank unsubscribe hmget del select psync wait host: multi xack zcount xgroup georadius rpush exists blpop zinterstore watch xinfo zrevrange ping hscan sismember lrange asking zscan ltrim flushdb module monitor readonly geopos xlen post memory randomkey time debug incrby strlen hset persist srandmember pfcount hincrbyfloat xadd geohash command sdiffstore xread type object brpoplpush zrangebylex expire hstrlen getrange lset replconf subscribe sunion xpending unlink save dump zrem get config psubscribe ttl sinterstore evalsha decrby pfadd lpos xdel pfselftest zscore rpoplpush llen pfmerge setbit hget setex restore-asking hello setrange stralgo incr restore pexpire scan bzpopmax sdiff spop pfdebug zrank hmset latency xrange hdel pexpireat zcard bzpopmin cluster zrevrangebylex lpush rpushx hkeys zrevrangebyscore set incrbyfloat renamenx xsetid publish hsetnx psetex mset msetnx punsubscribe pttl sunionstore hincrby zadd georadius_ro script readwrite bitpos touch rename zpopmax eval info expireat smembers zlexcount keys bitcount smove bgsave georadiusbymember_ro zpopmin migrate getset slowlog sort lastsave echo role
localhost:6379_0$ command help
./redis.bash: eval: line 104: syntax error near unexpected token `<'
./redis.bash: eval: line 104: `REDIS_REPLY=(COMMAND <subcommand> arg arg ... arg. Subcommands are: (no subcommand) -- Return details about all Redis commands. COUNT -- Return the total number of commands in this Redis server. GETKEYS <full-command> -- Return the keys from a full Redis command. INFO [command-name ...] -- Return details about multiple Redis commands.)'
localhost:6379_0$ 
```

... so maybe we shouldn't recommend this client just yet...

SomajitDey commented:
@zuiderkwast Thank you so much for pointing out the bugs 👍. I have fixed the line 104 error and some other stupid mistakes...would be pushing the same soon. `command help` works now. I realized that I forgot to take into account the nested RESP arrays, such as returned by `command`. For such compound responses, simple Bash arrays won't do (as they don't support nesting). Removed the "recommended" flag accordingly.

zuiderkwast commented:
@SomajitDey I also tried to start with a few newlines in the CLI. It is also some strange behaviour. How about we let this PR wait a bit until these problems are fixed? I can test it again when you post a comment here. OK?

SomajitDey commented:
@zuiderkwast Alrighty. I will keep you posted.

SomajitDey commented:
@zuiderkwast Hi again. I have taken care of the nested RESP arrays and `command` is working now. Can you kindly test my latest [push](https://github.com/SomajitDey/redis-client#acknowledgements).

[Please note that I have chosen not to print +OK and +PONG in the CLI as these are trivial- the CLI prompt returning without error is enough to understand the server has responded OK or with PONG (i.e. in case of `ping`).]

zuiderkwast commented:
OK, nested arrays work now. They're printed with indentation and leading dashes.

The CLI still breaks when I start with a newline. Then it seems to get stuck. Also, I think it's confusing to hide the PONG and OK results, since no other redis CLI does that. As a user, you could think that the command didn't get through. Not a big deal though. I think this can be merged.

Here's my test session:

```
$ ./redis-cli 
localhost:6379_0$ hset k foo bar baz quux
(int) 2
localhost:6379_0$ hgetall k
--foo
 -bar
 -baz
 -quux
localhost:6379_0$ set foo bar
localhost:6379_0$ ping
localhost:6379_0$ 
get foo



^C
```

SomajitDey commented:
@zuiderkwast 
Just fixed the newline bug...you did point it out earlier too but I missed it, my bad. Also OK and PING will now be printed in all terminal sessions. Thank you once again for making this client better. 

Here is my last session for reference:

```bash
somajit@DESKTOP-N5BP6IN:~/my-repos/redis-client$ ./redis-cli
localhost:6379_0$
localhost:6379_0$
localhost:6379_0$
localhost:6379_0$ ping
PONG
localhost:6379_0$ set foo bar
OK
localhost:6379_0$ q
somajit@DESKTOP-N5BP6IN:~/my-repos/redis-client$
```

SomajitDey commented:
@zuiderkwast 
Hi again. This client [now ](https://github.com/SomajitDey/redis-client)supports pub-sub [`(P)SUBSCRIBE` & `MONITOR` commands] and a functional connection-pool. The library-functions also support pipelining. Also,  now that you are satisfied that the earlier bugs have been fixed, do you think this client can be **recommended?**

**Pub/Sub example:** 
```
localhost:6379_0$ subscribe topic
--subscribe
 -topic
 -(int) 1
--message
 -topic
 -hello redis community
--message
 -topic
 -is this client still lacking?
^CPress Enter
localhost:6379_0$ get foo
bar
localhost:6379_0$
```

**Connection-pool and pipelining:**
```
somajit@DESKTOP-N5BP6IN:~/my-repos/redis-client$ ./redis-pool -n2 7000 &
[2] 5890
somajit@DESKTOP-N5BP6IN:~/my-repos/redis-client$ . ./redis.bash
somajit@DESKTOP-N5BP6IN:~/my-repos/redis-client$ echo -e 'SET foo bar\nGET foo' | nc -N localhost 7000 | redis_rep -n2
OK
bar
somajit@DESKTOP-N5BP6IN:~/my-repos/redis-client$
```

zuiderkwast commented:
Nice! Keep up the good work on the client @SomajitDey!
